src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToLinqMethodProvider.cs

```diff
@@ -0,0 +1,169 @@
+                if (selectExpression is IdentifierNameSyntax identifierName)
+                {
+                    // TODO consider a better condition to compare
+                    if (identifierName.Identifier.Text == identifier.Text)
+                    {
+                        return expression;
+                    }
+                }
+
+                return CreateInvocationExpression(expression, identifier, selectClause.Expression, "Select");
+            }
+
+            private static ExpressionSyntax CreateInvocationExpression(ExpressionSyntax expression, SyntaxToken identifier, ExpressionSyntax bodyExpression, string keyword)
+            {
+                var parameter = SyntaxFactory.Parameter(identifier);
+                var lambda = SyntaxFactory.SimpleLambdaExpression(parameter, bodyExpression.WithoutTrailingTrivia());
+                var argument = SyntaxFactory.Argument(lambda);
+                var arguments = new SeparatedSyntaxList<ArgumentSyntax>().Add(argument);
+
+                return CreateInvocationExpression(expression, arguments, keyword);
+            }
+
+            private static InvocationExpressionSyntax CreateInvocationExpression(ExpressionSyntax expression, SeparatedSyntaxList<ArgumentSyntax> arguments, string keyword)
+            {
+                var argumentList = SyntaxFactory.ArgumentList(arguments);
+                var identifierName = SyntaxFactory.IdentifierName(keyword);
+                var memberAccessExpression = SyntaxFactory.MemberAccessExpression(SyntaxKind.SimpleMemberAccessExpression, expression, identifierName);
+                return SyntaxFactory.InvocationExpression(memberAccessExpression, argumentList);
+            }
+
+            private static ExpressionSyntax ProcessOrderByClause(ExpressionSyntax expression, SyntaxToken identifier, OrderByClauseSyntax orderByClause)
+            {
+                bool isFirst = true;
+                foreach (var ordering in orderByClause.Orderings)
+                {
+                    string orderingKeyword;
+                    switch (ordering.Kind())
+                    {
+                        case SyntaxKind.AscendingOrdering: orderingKeyword = isFirst ? "OrderBy" : "ThenBy"; break;
+                        case SyntaxKind.DescendingOrdering: orderingKeyword = isFirst ? "OrderByDescending" : "ThenByDescending"; break;
+                        default: return null;
+                    }
+                    expression = CreateInvocationExpression(expression, identifier, ordering.Expression, orderingKeyword);
+                    isFirst = false;
+                }
+
+                return expression;
+            }
+
+            protected override QueryExpressionSyntax FindNodeToRefactor(SyntaxNode root, CodeRefactoringContext context)
+            {
+                return root.FindNode(context.Span).FirstAncestorOrSelf<QueryExpressionSyntax>();
+            }
+
+            private class ConversionContext
```

Comment:
1. Is this class needed?  
2. Can you use a struct instead? 
3. Would a tuple be ok?

Reply:
I think this class will be extended to something like 
[https://github.com/dotnet/roslyn/blob/d4dab355b96955aca5b4b0ebf6282575fad78ba8/src/Compilers/CSharp/Portable/Binder/Binder.QueryTranslationState.cs](https://github.com/dotnet/roslyn/blob/d4dab355b96955aca5b4b0ebf6282575fad78ba8/src/Compilers/CSharp/Portable/Binder/Binder.QueryTranslationState.cs)

---
In reply to: [167387799](https://github.com/dotnet/roslyn/pull/24759#discussion_r167387799) [](ancestors = 167387799)